File: scripts/generate.ts
import 'dotenv/config';
import axios from 'axios';
import { GoogleGenAI } from '@google/genai';
import fs from 'fs';
import path from 'path';
import nodemailer from 'nodemailer';

// ================= CONFIG =================
const EXTERNAL_API_BASE = 'https://cloud-text-manager-server.vercel.app';
const EXTERNAL_API_URL = `${EXTERNAL_API_BASE}/api/all-files`;
const GENERATED_DIR = path.join(process.cwd(), 'generated-content');

const BATCH_SIZE = 2;
const CONCURRENCY_LIMIT = 1;
const REQUEST_DELAY = 1500;
const MAX_RETRIES = 5;

// Ensure output directory exists
if (!fs.existsSync(GENERATED_DIR)) {
  fs.mkdirSync(GENERATED_DIR, { recursive: true });
}

// ================= GEMINI MULTI-KEY =================
const apiKeys = Object.keys(process.env)
  .filter((key) => key.startsWith('GEMINI_API_KEY'))
  .sort()
  .map((key) => process.env[key])
  .filter(Boolean) as string[];

if (apiKeys.length === 0) {
  throw new Error('No GEMINI_API_KEY variables found.');
}

console.log(`Loaded ${apiKeys.length} Gemini API keys.`);

let keyIndex = 0;
const getNextGemini = () => {
  const key = apiKeys[keyIndex];
  keyIndex = (keyIndex + 1) % apiKeys.length;
  return new GoogleGenAI({ apiKey: key });
};

// ================= SLUG GENERATOR =================
const generateSafeFilename = (originalFilename: string): string => {
  const baseName = path.parse(originalFilename).name;
  return (
    baseName
      .toLowerCase()
      .normalize('NFKD')
      .replace(/[^\w\s-]/g, '')
      .trim()
      .replace(/\s+/g, '-')
      .replace(/-+/g, '-')
      .replace(/^-|-$/g, '') + '.mdx'
  );
};

// ================= EMAIL =================
const sendEmail = async (subject: string, text: string) => {
  if (!process.env.EMAIL_HOST) return;

  const transporter = nodemailer.createTransport({
    host: process.env.EMAIL_HOST,
    port: Number(process.env.EMAIL_PORT) || 587,
    secure: process.env.EMAIL_SECURE === 'true',
    auth: {
      user: process.env.EMAIL_USER,
      pass: process.env.EMAIL_PASS,
    },
  });

  await transporter.sendMail({
    from: process.env.EMAIL_FROM || process.env.EMAIL_USER,
    to: process.env.EMAIL_TO || process.env.EMAIL_USER,
    subject,
    text,
  });
};

// ================= UTILS =================
const sleep = (ms: number) => new Promise((res) => setTimeout(res, ms));

// ================= RETRY + KEY ROTATION =================
const generateWithRetry = async (promptText: string, model: string) => {
  let attempt = 0;

  while (attempt < MAX_RETRIES) {
    try {
      const ai = getNextGemini();
      await sleep(REQUEST_DELAY);

      const result = await ai.models.generateContent({
        model,
        contents: promptText,
        config: { responseMimeType: 'text/plain' },
      });

      if (!result.text) throw new Error('Empty response');
      return result.text;

    } catch (error: any) {
      attempt++;
      console.log(`Retry ${attempt}/${MAX_RETRIES}`);

      if (error?.response?.status === 429) {
        console.log('Rate limit hit → switching key');
      }

      await sleep(1000 * attempt);

      if (attempt >= MAX_RETRIES) throw error;
    }
  }
};

// ================= FILE PROCESS =================
const processFile = async (file: any, model: string) => {
  console.log(`Processing: ${file.originalFilename}`);

  try {
    const promptResponse = await axios.get(file.secureUrl);
    const promptText =
      typeof promptResponse.data === 'string'
        ? promptResponse.data
        : JSON.stringify(promptResponse.data);

    const generatedContent = await generateWithRetry(promptText, model);

    const safeFilename = generateSafeFilename(file.originalFilename);
    const filePath = path.join(GENERATED_DIR, safeFilename);

    fs.writeFileSync(filePath, generatedContent, 'utf-8');

    await axios.put(`${EXTERNAL_API_URL}/${file._id}`, {
      status: 'AlreadyCopy',
      completedTimestamp: Date.now(),
    });

    console.log(`✅ Done: ${safeFilename}`);
    return { success: true, name: safeFilename };

  } catch (err: any) {
    console.error(`❌ Failed: ${file.originalFilename}`, err.message);
    return { success: false, name: file.originalFilename, error: err.message };
  }
};

// ================= PARALLEL QUEUE =================
const runQueue = async (files: any[]) => {
  const model = process.env.GEMINI_MODEL || 'gemini-3-flash-preview';
  const results: any[] = [];
  let index = 0;

  async function worker() {
    while (true) {
      if (index >= files.length) break;
      const currentIndex = index++;
      const file = files[currentIndex];
      const result = await processFile(file, model);
      results.push(result);
    }
  }

  const workers = Array.from({ length: CONCURRENCY_LIMIT }, () => worker());
  await Promise.all(workers);

  return results;
};

// ================= FOLDER STATS =================
const getFolderStats = (dir: string) => {
  if (!fs.existsSync(dir)) return { totalFiles: 0, mdxFiles: 0 };
  const files = fs.readdirSync(dir);
  const mdxFiles = files.filter((f) => f.endsWith('.mdx')).length;
  return { totalFiles: files.length, mdxFiles };
};

// ================= SEND BATCH EMAIL =================
const sendBatchEmail = async (success: any[], failed: any[]) => {
  // Bangladesh time
  const now = new Date();
  const dateTimeStr = now.toLocaleString('en-US', { timeZone: 'Asia/Dhaka' });

  const folderStats = getFolderStats(GENERATED_DIR);

  const subject = `Content Generation Report lawn - ${dateTimeStr} - ${success.length} Success, ${failed.length} Failed`;

  const body = `
Multi-Key Content Generation Report
----------------------------------
Date & Time (BDT): ${dateTimeStr}
Total Processed: ${success.length + failed.length}
Success: ${success.length}
Failed: ${failed.length}

Successful Files:
${success.map((f) => `- ${f.name}`).join('\n')}

Failed Files:
${failed.map((f) => `- ${f.name} (${f.error})`).join('\n')}

Folder Stats:
- Total files in folder: ${folderStats.totalFiles}
- Total MDX files: ${folderStats.mdxFiles}

Generated content has been saved to the repository.
`;

  await sendEmail(subject, body);
};

// ================= MAIN =================
const run = async () => {
  console.log('🚀 Starting Multi-Key Parallel Processing...');

  try {
    const response = await axios.get("https://cloud-text-manager-server.vercel.app/api/all-files?section=General");
    const pendingFiles = response.data
      .filter((f: any) => f.status === 'Pending')
      .slice(0, BATCH_SIZE);

    if (!pendingFiles.length) {
      console.log('No pending files.');
      return;
    }

    console.log(`Found ${pendingFiles.length} files.`);

    const results = await runQueue(pendingFiles);

    const success = results.filter((r) => r.success);
    const failed = results.filter((r) => !r.success);

    await sendBatchEmail(success, failed);

    console.log('🎉 Batch Completed Successfully');

  } catch (error: any) {
    console.error('Fatal error:', error.message);
    await sendEmail('Batch Failed', error.message);
    process.exit(1);
  }
};


run();

File: scripts/generate.py
import os
import sys
import time
import json
import smtplib
from email.mime.text import MIMEText
import requests
from dotenv import load_dotenv
import re
from datetime import datetime, timezone, timedelta
from concurrent.futures import ThreadPoolExecutor, as_completed

load_dotenv()

# ================= CONFIG =================
EXTERNAL_API_BASE = 'https://cloud-text-manager-server.vercel.app'
EXTERNAL_API_URL = f'{EXTERNAL_API_BASE}/api/all-files'
GENERATED_DIR = os.path.join(os.getcwd(), 'generated-content')

BATCH_SIZE = 1
CONCURRENCY_LIMIT = 1
REQUEST_DELAY = 1.5
MAX_RETRIES = 5

os.makedirs(GENERATED_DIR, exist_ok=True)

# ================= OPENROUTER =================
OPENROUTER_API_KEY = os.environ.get('OPENROUTER_API_KEY')
if not OPENROUTER_API_KEY:
    raise ValueError('No OPENROUTER_API_KEY variable found.')

def generate_response(messages):
    url = "https://openrouter.ai/api/v1/chat/completions"
    headers = {
        "Authorization": f"Bearer {OPENROUTER_API_KEY}",
        "Content-Type": "application/json"
    }
    data = {
        "model": "z-ai/glm-4.5-air:free",
        "messages": messages
    }
    
    response = requests.post(url, headers=headers, json=data)
    response.raise_for_status()
    return response.json()

# ================= SLUG GENERATOR =================
def generate_safe_filename(original_filename):
    base_name = os.path.splitext(original_filename)[0]
    # To mimic JS replace logic: string to lower, remove non-word chars, trim, replace spaces/dashes with single dash
    slug = base_name.lower()
    slug = re.sub(r'[^\w\s-]', '', slug)
    slug = slug.strip()
    slug = re.sub(r'\s+', '-', slug)
    slug = re.sub(r'-+', '-', slug)
    slug = slug.strip('-')
    return slug + '.mdx'

# ================= EMAIL =================
def send_email(subject, text):
    email_host = os.environ.get('EMAIL_HOST')
    if not email_host:
        return
        
    email_port = int(os.environ.get('EMAIL_PORT', 587))
    email_secure = os.environ.get('EMAIL_SECURE', '').lower() == 'true'
    email_user = os.environ.get('EMAIL_USER')
    email_pass = os.environ.get('EMAIL_PASS')
    email_from = os.environ.get('EMAIL_FROM', email_user)
    email_to = os.environ.get('EMAIL_TO', email_user)
    
    msg = MIMEText(text)
    msg['Subject'] = subject
    msg['From'] = email_from
    msg['To'] = email_to
    
    try:
        if email_secure or email_port == 465:
            server = smtplib.SMTP_SSL(email_host, email_port)
        else:
            server = smtplib.SMTP(email_host, email_port)
            server.starttls()
            
        server.login(email_user, email_pass)
        server.send_message(msg)
        server.quit()
    except Exception as e:
        print(f"Failed to send email: {e}")

# ================= UTILS =================
def sleep(seconds):
    time.sleep(seconds)

# ================= RETRY LOGIC =================
def generate_with_retry(prompt_text):
    attempt = 0
    
    while attempt < MAX_RETRIES:
        try:
            sleep(REQUEST_DELAY)
            
            messages = [
                {
                    "role": "user",
                    "content": prompt_text
                }
            ]
            response_data = generate_response(messages)
            
            choices = response_data.get('choices', [])
            if not choices:
                raise ValueError('Empty response')
                
            content = choices[0].get('message', {}).get('content')
            if not content:
                raise ValueError('Empty response content')
                
            return content
            
        except Exception as e:
            attempt += 1
            print(f"Retry {attempt}/{MAX_RETRIES}")
            
            # Simple rate limit handling
            if hasattr(e, 'response') and e.response is not None and getattr(e.response, 'status_code', None) == 429:
                print('Rate limit hit')
                
            sleep(1.0 * attempt)
            
            if attempt >= MAX_RETRIES:
                raise e

# ================= FILE PROCESS =================
def process_file(file_data):
    original_filename = file_data.get('originalFilename')
    print(f"Processing: {original_filename}")
    
    try:
        req = requests.get(file_data.get('secureUrl'))
        req.raise_for_status()
        
        try:
            prompt_text = req.json()
            if not isinstance(prompt_text, str):
                prompt_text = json.dumps(prompt_text)
        except ValueError:
            prompt_text = req.text
            
        generated_content = generate_with_retry(prompt_text)
        
        safe_filename = generate_safe_filename(original_filename)
        file_path = os.path.join(GENERATED_DIR, safe_filename)
        
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(generated_content)
            
        put_url = f"{EXTERNAL_API_URL}/{file_data.get('_id')}"
        put_data = {
            "status": "AlreadyCopy",
            "completedTimestamp": int(time.time() * 1000)
        }
        requests.put(put_url, json=put_data)
        
        print(f"✅ Done: {safe_filename}")
        return {"success": True, "name": safe_filename}
        
    except Exception as e:
        print(f"❌ Failed: {original_filename} - {str(e)}")
        return {"success": False, "name": original_filename, "error": str(e)}

# ================= PARALLEL QUEUE =================
def run_queue(files):
    results = []
    # If CONCURRENCY_LIMIT is 1, regular loop is fine, but we mimic JS Promise.all style
    with ThreadPoolExecutor(max_workers=max(1, CONCURRENCY_LIMIT)) as executor:
        future_to_file = {executor.submit(process_file, f): f for f in files}
        for future in as_completed(future_to_file):
            results.append(future.result())
    return results

# ================= FOLDER STATS =================
def get_folder_stats(directory):
    if not os.path.exists(directory):
        return {"totalFiles": 0, "mdxFiles": 0}
        
    files = os.listdir(directory)
    mdx_files = sum(1 for f in files if f.endswith('.mdx'))
    return {"totalFiles": len(files), "mdxFiles": mdx_files}

# ================= SEND BATCH EMAIL =================
def send_batch_email(success, failed):
    tz = timezone(timedelta(hours=6)) # Dhaka is UTC+6
    now = datetime.now(tz)
    date_time_str = now.strftime('%m/%d/%Y, %I:%M:%S %p')
    
    folder_stats = get_folder_stats(GENERATED_DIR)
    
    subject = f"Content Generation Report lawn - {date_time_str} - {len(success)} Success, {len(failed)} Failed"
    
    success_list = '\n'.join([f"- {f['name']}" for f in success])
    failed_list = '\n'.join([f"- {f['name']} ({f.get('error')})" for f in failed])
    
    body = f'''Multi-Key Content Generation Report
----------------------------------
Date & Time (BDT): {date_time_str}
Total Processed: {len(success) + len(failed)}
Success: {len(success)}
Failed: {len(failed)}

Successful Files:
{success_list}

Failed Files:
{failed_list}

Folder Stats:
- Total files in folder: {folder_stats["totalFiles"]}
- Total MDX files: {folder_stats["mdxFiles"]}

Generated content has been saved to the repository.
'''
    send_email(subject, body)

# ================= MAIN =================
def main():
    print('🚀 Starting Processing...')
    
    try:
        response = requests.get(f"{EXTERNAL_API_BASE}/api/all-files?section=General")
        response.raise_for_status()
        
        all_files = response.json()
        pending_files = [f for f in all_files if f.get('status') == 'Pending'][:BATCH_SIZE]
        
        if not pending_files:
            print('No pending files.')
            return
            
        print(f"Found {len(pending_files)} files.")
        
        results = run_queue(pending_files)
        
        success = [r for r in results if r['success']]
        failed = [r for r in results if not r['success']]
        
        send_batch_email(success, failed)
        
        print('🎉 Batch Completed Successfully')
        
    except Exception as e:
        print(f'Fatal error: {str(e)}')
        send_email('Batch Failed', str(e))
        sys.exit(1)

if __name__ == "__main__":
    main()
